libraries/ms-common/src/stm32f0xx/adc.c

```diff
@@ -0,0 +1,208 @@
+
+  // Calculate the ADC calibration factor
+  ADC_GetCalibrationFactor(ADC1);
+
+  ADC_TempSensorCmd(ENABLE);
+  ADC_VrefintCmd(ENABLE);
+
+  ADC_ContinuousModeCmd(ADC1, adc_mode);
+  while (ADC_GetFlagStatus(ADC1, ADC_FLAG_ADCAL)) { }
+
+  // Enable the ADC
+  ADC_Cmd(ADC1, ENABLE);
+  while (!ADC_GetFlagStatus(ADC1, ADC_FLAG_ADRDY)) { }
+
+  // Set the ADC to wait mode
+  ADC_WaitModeCmd(ADC1, ENABLE);
+  ADC_AutoPowerOffCmd(ADC1, ENABLE);
+
+  // Enable interrupts for the end of each conversion
+  stm32f0xx_interrupt_nvic_enable(ADC1_COMP_IRQn, INTERRUPT_PRIORITY_NORMAL);
+  ADC_ITConfig(ADC1, ADC_IER_EOCIE, ENABLE);
+
+  // Initialize static varables
+  s_adc_status.continuous = adc_mode;
+  s_adc_status.current_channel = 0;
+
+  // Start background conversions if in continuous
+  if (adc_mode) {
+    ADC_StartOfConversion(ADC1);
+  }
+}
+
+StatusCode adc_set_channel(ADCChannel adc_channel, bool new_state) {
+  if (adc_channel >= NUM_ADC_CHANNEL) {
+    return STATUS_CODE_INVALID_ARGS;
+  }
+
+  if (new_state) {
+    ADC_ChannelConfig(ADC1, (1 << adc_channel), ADC_SampleTime_239_5Cycles);
+  } else {
+    ADC1->CHSELR &= ~(1 << adc_channel);
+  }
+
+  // The bridge divider is only enabled when needed to minimize consumption on the battery
+  if (adc_channel == ADC_CHANNEL_BAT) {
+    ADC_VbatCmd(new_state);
+  }
+
+  s_adc_status.select = ADC1->CHSELR;
+
+  return STATUS_CODE_OK;
+}
+
+StatusCode adc_register_callback(ADCChannel adc_channel, ADCCallback callback, void *context) {
+  // Returns invalid if the given address is not connected to an ADC channel
+  if (adc_channel >= NUM_ADC_CHANNEL) {
+    return STATUS_CODE_INVALID_ARGS;
+  }
+
+  s_adc_interrupts[adc_channel].callback = callback;
+  s_adc_interrupts[adc_channel].context = context;
+
+  return STATUS_CODE_OK;
+}
+
+StatusCode adc_read_value(ADCChannel adc_channel, uint16_t *reading) {
+  if (adc_channel >= NUM_ADC_CHANNEL) {
+    return STATUS_CODE_INVALID_ARGS;
+  }
+
+  if (!s_adc_status.continuous) {
+    s_adc_status.current_channel = 0;
+    ADC_StartOfConversion(ADC1);
+    while (!ADC_GetFlagStatus(ADC1, ADC_FLAG_EOSEQ)) { }
+  }
+
+  reading = s_adc_interrupts[adc_channel].reading;
+
+  return STATUS_CODE_OK;
+}
+
+
+StatusCode adc_trigger_callback(ADCChannel adc_channel) {
+  if (adc_channel >= NUM_ADC_CHANNEL) {
+    return STATUS_CODE_INVALID_ARGS;
+  }
+
+  ADC_StopOfConversion(ADC1);
+  while (ADC_GetFlagStatus(ADC1, ADC_FLAG_ADSTP)) { }
+
+  // Save the current channel selection and set only the given channel
+  uint32_t temp = ADC1->CHSELR;
+  ADC1->CHSELR &= (1 << adc_channel);
+
+  s_adc_status.current_channel = 0;
+  ADC_StartOfConversion(ADC1);
+  while (!ADC_GetFlagStatus(ADC1, ADC_FLAG_EOSEQ)) { }
+
+  ADC1->CHSELR = temp;
+
+  return STATUS_CODE_OK;
+}
+
+
+void ADC1_COMP_IRQHandler() {
+  ADCChannel current_channel = s_adc_status.current_channel;
+  uint32_t select = s_adc_status.select;
+  uint16_t reading;
+
+  if (select & (1 << current_channel)) {
```

Comment:
If the interrupt is called, I think we can assume that your current channel is valid.